This file is part of MXE. See LICENSE.md for licensing information.

Contains ad hoc patches for cross building.

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MXE <mxe@example.invalid>
Date: Sat, 28 Mar 2026 00:00:00 +0000
Subject: [PATCH] Backport build system fixes

Backport the downstream USDX packaging fixes so projectM installs the
public headers needed by consumers, avoids requiring generated config.h
from Common.hpp, fixes DESTDIR handling for preset installation,
restores pkg-config metadata generation, and links libprojectM against
the detected dlfcn library.

diff --git a/Makefile.am b/Makefile.am
--- a/Makefile.am
+++ b/Makefile.am
@@ -18,8 +18,8 @@ pm_shaders__DATA = src/libprojectM/Renderer/blur.cg \
 
 # find and install all preset files
 install-data-local:
-	test -z $(pkgdatadir) || $(MKDIR_P) $(pm_presets_dir)
-	find "$(PRESETSDIR)" -type f -exec $(INSTALL_DATA) {} $(pm_presets_dir) \;
+	test -z $(DESTDIR)$(pkgdatadir) || $(MKDIR_P) $(DESTDIR)$(pm_presets_dir)
+	find "$(PRESETSDIR)" -type f -exec $(INSTALL_DATA) {} $(DESTDIR)$(pm_presets_dir) \;
 
 # from https://stackoverflow.com/questions/30897170/ac-subst-does-not-expand-variable answer: https://stackoverflow.com/a/30960268
 # ptomato https://stackoverflow.com/users/172999/ptomato
diff --git a/autogen.sh b/autogen.sh
new file mode 100755
--- /dev/null
+++ b/autogen.sh
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+autoreconf --install || exit 1
+
+echo "Now run ./configure && make"
diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -11,3 +11,4 @@ AX_CHECK_GL
 AC_CHECK_LIB(c, dlopen, LIBDL="", AC_CHECK_LIB(dl, dlopen, LIBDL="-ldl"))
+AC_SUBST([LIBDL])
 
 AC_CONFIG_HEADERS([config.h])
@@ -19,8 +20,9 @@ AC_CONFIG_FILES([
   src/libprojectM/Renderer/Makefile
   src/libprojectM/NativePresetFactory/Makefile
   src/libprojectM/MilkdropPresetFactory/Makefile
+  src/libprojectM/libprojectM.pc
   src/NativePresets/Makefile
   src/projectM-sdl/Makefile
   src/projectM-qt/Makefile
   src/projectM-pulseaudio/Makefile
 ])

@@ -91,6 +93,7 @@ case $host_os in
   darwin*)
     # OSX needs CoreFoundation
     AC_MSG_RESULT(Apple hoarderware detected)
+    NATIVE_PRESETS_SUBDIR=NativePresets
     LIBS="$LIBS -framework CoreFoundation"
     ;;
   linux*)
@@ -98,10 +101,17 @@ case $host_os in
     # limux needs dl
     AC_MSG_RESULT(GNU/LINUX detected)
     LIBS="$LIBS -ldl"
+    NATIVE_PRESETS_SUBDIR=NativePresets
+    ;;
+  mingw* | msys*)
+    AC_MSG_RESULT(MinGW detected)
+    NATIVE_PRESETS_SUBDIR=
+    ;;
+  *)
+    NATIVE_PRESETS_SUBDIR=NativePresets
     ;;
 esac
-
-
+AC_SUBST([NATIVE_PRESETS_SUBDIR])
 

diff --git a/src/Makefile.am b/src/Makefile.am
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -9 +9 @@ endif
-SUBDIRS=libprojectM NativePresets ${PROJECTM_SDL_SUBDIR} ${PROJECTM_QT_SUBDIR}
+SUBDIRS=libprojectM ${NATIVE_PRESETS_SUBDIR} ${PROJECTM_SDL_SUBDIR} ${PROJECTM_QT_SUBDIR}
@@ -9,0 +10,4 @@ SUBDIRS=libprojectM ${NATIVE_PRESETS_SUBDIR} ${PROJECTM_SDL_SUBDIR} ${PROJECTM_QT_SUBDIR}
+
+# system headers/libraries/data to install
+# for compatibility reasons here as nobase_include
+nobase_include_HEADERS = libprojectM/projectM.hpp libprojectM/Common.hpp libprojectM/dlldefs.h libprojectM/event.h libprojectM/fatal.h libprojectM/PCM.hpp
diff --git a/src/NativePresets/Makefile.am b/src/NativePresets/Makefile.am
--- a/src/NativePresets/Makefile.am
+++ b/src/NativePresets/Makefile.am
@@ -17,6 +17,15 @@ presets_LTLIBRARIES = \
   libRovastarFractalSpiral.la \
   libRovastarFractopiaFrantic.la
 
+AM_LDFLAGS = -module -avoid-version -no-undefined
+
+libMstressJuppyDancer_la_LIBADD = $(top_builddir)/src/libprojectM/libprojectM.la
+libRLGFractalDrop7c_la_LIBADD = $(top_builddir)/src/libprojectM/libprojectM.la
+libRovastarDarkSecret_la_LIBADD = $(top_builddir)/src/libprojectM/libprojectM.la
+libRovastarDriftingChaos_la_LIBADD = $(top_builddir)/src/libprojectM/libprojectM.la
+libRovastarFractalSpiral_la_LIBADD = $(top_builddir)/src/libprojectM/libprojectM.la
+libRovastarFractopiaFrantic_la_LIBADD = $(top_builddir)/src/libprojectM/libprojectM.la
+
 libMstressJuppyDancer_la_SOURCES = MstressJuppyDancer.cpp
 libRLGFractalDrop7c_la_SOURCES = RLGFractalDrop7c.cpp
 libRovastarDarkSecret_la_SOURCES = RovastarDarkSecret.cpp
diff --git a/src/libprojectM/Common.hpp b/src/libprojectM/Common.hpp
--- a/src/libprojectM/Common.hpp
+++ b/src/libprojectM/Common.hpp
@@ -30,7 +30,6 @@
 #include <typeinfo>
 #include <cstdarg>
 #include <cassert>
-#include "config.h"
 #ifdef _MSC_sVER
 #define strcasecmp(s, t) _strcmpi(s, t)
 #endif
diff --git a/src/libprojectM/Makefile.am b/src/libprojectM/Makefile.am
--- a/src/libprojectM/Makefile.am
+++ b/src/libprojectM/Makefile.am
@@ -10,8 +10,6 @@ AM_CPPFLAGS = \
 	-I$(top_srcdir)/src/libprojectM/Renderer \
 	${FTGL_CFLAGS}
 
-# system headers/libraries/data to install
-include_HEADERS = projectM.hpp
 lib_LTLIBRARIES = libprojectM.la  # public, possibly-shared library
 
 # link flags
@@ -16,7 +14,8 @@ libprojectM_la_LDFLAGS = -no-undefined -version-info 0:2:0
 # link libRenderer, MilkdropPresetFactory, NativePresetFactory, and libprojectM sources
 libprojectM_la_LIBADD = \
 $(top_srcdir)/src/libprojectM/MilkdropPresetFactory/libMilkdropPresetFactory.la \
-$(top_srcdir)/src/libprojectM/NativePresetFactory/libNativePresetFactory.la \
+$(top_srcdir)/src/libprojectM/NativePresetFactory/libNativePresetFactory.la \
+$(LIBDL) \
 $(top_srcdir)/src/libprojectM/Renderer/libRenderer.la
 libprojectM_la_SOURCES = ConfigFile.cpp Preset.cpp PresetLoader.cpp timer.cpp \
   KeyHandler.cpp PresetChooser.cpp TimeKeeper.cpp PCM.cpp PresetFactory.cpp \
@@ -47,6 +46,6 @@ libprojectM_la_SOURCES = ConfigFile.cpp Preset.cpp PresetLoader.cpp timer.cpp \
 	omptl/omptl_algorithm
 
 pkgconfigdir = $(libdir)/pkgconfig
-# pkgconfig_DATA = src/libprojectM.pc
-# EXTRA_DIST += src/libprojectM.pc.in
-# CLEANFILES += src/libprojectM.pc
+pkgconfig_DATA = libprojectM.pc
+EXTRA_DIST += libprojectM.pc.in
+CLEANFILES += libprojectM.pc
diff --git a/src/libprojectM/libprojectM.pc.in b/src/libprojectM/libprojectM.pc.in
new file mode 100644
--- /dev/null
+++ b/src/libprojectM/libprojectM.pc.in
@@ -0,0 +1,13 @@
+prefix=@prefix@
+exec_prefix=@exec_prefix@
+libdir=@libdir@
+includedir=@includedir@
+pkgdatadir=@datadir@/@PACKAGE_NAME@
+sysconfdir=@datadir@/@PACKAGE_NAME@
+
+Name: libprojectM
+Version: @PACKAGE_VERSION@
+Description: projectM - OpenGL Milkdrop
+Requires:
+Libs: -L${libdir} -lprojectM
+Cflags: -I${includedir}
